// @code-hike/highlighter/dist/index.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var shiki = require('shiki');

/*! *****************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */

var __assign = function() {
    __assign = Object.assign || function __assign(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};

function __awaiter(thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
}

function __generator(thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
}

var _a;
var highlighterPromise = null;
var highlighter = null;
var newlineRe = /\r\n|\r|\n/;
function highlight(_a) {
    var code = _a.code, lang = _a.lang, theme = _a.theme;
    return __awaiter(this, void 0, void 0, function () {
        var lines_1, tokenizedLines, lines;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    if (lang === "text") {
                        lines_1 = code ? code.split(newlineRe) : [""];
                        return [2 /*return*/, {
                                lines: lines_1.map(function (line) { return ({
                                    tokens: [{ content: line, props: {} }],
                                }); }),
                                lang: lang,
                            }];
                    }
                    if (highlighterPromise === null) {
                        shiki.setCDN("https://unpkg.com/shiki/");
                        highlighterPromise = shiki.getHighlighter({
                            theme: theme,
                        });
                    }
                    if (!(highlighter === null)) return [3 /*break*/, 2];
                    return [4 /*yield*/, highlighterPromise];
                case 1:
                    highlighter = _b.sent();
                    _b.label = 2;
                case 2:
                    if (!missingTheme(highlighter, theme)) return [3 /*break*/, 4];
                    return [4 /*yield*/, highlighter.loadTheme(theme)];
                case 3:
                    _b.sent();
                    _b.label = 4;
                case 4:
                    if (!missingLang(highlighter, lang)) return [3 /*break*/, 8];
                    _b.label = 5;
                case 5:
                    _b.trys.push([5, 7, , 8]);
                    return [4 /*yield*/, highlighter.loadLanguage(lang)];
                case 6:
                    _b.sent();
                    return [3 /*break*/, 8];
                case 7:
                    _b.sent();
                    console.warn("[Code Hike warning]", lang + " is not a valid language, no syntax highlighting will be applied.");
                    return [2 /*return*/, highlight({ code: code, lang: "text", theme: theme })];
                case 8:
                    tokenizedLines = highlighter.codeToThemedTokens(code, lang, theme.name, {
                        includeExplanation: false,
                    });
                    lines = tokenizedLines.map(function (line) { return ({
                        tokens: line.map(function (token) { return ({
                            content: token.content,
                            props: { style: getStyle(token) },
                        }); }),
                    }); });
                    return [2 /*return*/, { lines: lines, lang: lang }];
            }
        });
    });
}
function missingTheme(highlighter, theme) {
    return !highlighter
        .getLoadedThemes()
        .some(function (t) { return t === theme.name; });
}
function missingLang(highlighter, lang) {
    return !highlighter
        .getLoadedLanguages()
        .some(function (l) { return l === lang; });
}
var FONT_STYLE_TO_CSS = (_a = {},
    _a[shiki.FontStyle.NotSet] = {},
    _a[shiki.FontStyle.None] = {},
    _a[shiki.FontStyle.Italic] = { fontStyle: "italic" },
    _a[shiki.FontStyle.Bold] = { fontWeight: "bold" },
    _a[shiki.FontStyle.Underline] = { textDecoration: "underline" },
    _a);
function getStyle(token) {
    var fontStyle = token.fontStyle
        ? FONT_STYLE_TO_CSS[token.fontStyle]
        : {};
    return __assign({ color: token.color }, fontStyle);
}

exports.highlight = highlight;
