// @code-hike/mini-frame/dist/index.esm.js
import React from 'react';
import { useClasser, ClasserProvider } from '@code-hike/classer';
import { getColor, ColorName } from '@code-hike/utils';

/*! *****************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */

var __assign = function() {
    __assign = Object.assign || function __assign(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};

function __rest(s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
}

var MiniFrame = React.forwardRef(function (_a, ref) {
    var title = _a.title, children = _a.children, titleBar = _a.titleBar, classes = _a.classes, theme = _a.theme, _b = _a.zoom, zoom = _b === void 0 ? 1 : _b, overflow = _a.overflow, props = __rest(_a, ["title", "children", "titleBar", "classes", "theme", "zoom", "overflow"]);
    var c = useClasser("ch-frame", classes);
    var bar = titleBar || React.createElement(DefaultTitleBar, { title: title });
    var zoomStyle = {
        "--ch-frame-zoom": zoom,
        overflow: overflow,
    };
    return (React.createElement(ClasserProvider, { classes: classes },
        React.createElement("div", __assign({}, props, { ref: ref }),
            React.createElement("div", { className: c("") },
                React.createElement("div", { className: c("title-bar"), style: {
                        background: getColor(theme, ColorName.EditorGroupHeaderBackground),
                    } }, bar),
                React.createElement("div", { className: c("content") },
                    React.createElement("div", { className: c("zoom"), style: zoomStyle }, children))))));
});
var SimpleFrame = React.forwardRef(function (_a, ref) {
    var title = _a.title, children = _a.children, titleBar = _a.titleBar, classes = _a.classes; _a.overflow; var props = __rest(_a, ["title", "children", "titleBar", "classes", "overflow"]);
    var c = useClasser("ch", classes);
    var bar = titleBar || React.createElement(DefaultTitleBar, { title: title });
    return (React.createElement(ClasserProvider, { classes: classes },
        React.createElement("div", __assign({}, props, { ref: ref }),
            React.createElement("div", { className: c("simple-frame") },
                React.createElement("div", { className: c("frame-title-bar") }, bar),
                children))));
});
function DefaultTitleBar(_a) {
    var title = _a.title;
    var c = useClasser("ch-frame");
    return (React.createElement(React.Fragment, null,
        React.createElement("div", { className: c("left-bar") },
            React.createElement(FrameButtons, null)),
        React.createElement("div", { className: c("middle-bar") }, title),
        React.createElement("div", { className: c("right-bar") })));
}
function FrameButtons() {
    var c = useClasser("ch-frame");
    return (React.createElement("div", { className: c("buttons") },
        React.createElement("div", { className: c("button", "button-left") }),
        React.createElement("div", { className: c("button-space") }),
        React.createElement("div", { className: c("button", "button-middle") }),
        React.createElement("div", { className: c("button-space") }),
        React.createElement("div", { className: c("button", "button-right") })));
}

export { FrameButtons, MiniFrame, SimpleFrame };
